### DISCUSSION TITLE
Is it possible to use the renderer plugin in shortcodes?

### DISCUSSION THREAD

**Discussion by AleksandrHovhannisyan** [11ty/eleventy#2252] [Q&A] [answered]
**TL;DR**: In 11ty, when you create a shortcode using function declarations, they bind `this` to an object containing data about the page that invoked the shortcode. I'm wondering if it's possible to also expose other APIs, like `renderTemplate`. This would allow using those 11ty APIs from within a JavaScript shortcode.

I've registered `EleventyPluginRender` as shown in the docs, and I was wondering if it's possible to somehow access `renderTemplate` from within a JavaScript shortcode that uses a function declaration. If not, is this something that could be considered?

Use case: As an alternative to the approach described here, I could just do something like `this.renderTemplate(content, "md")` inside my shortcode and it would automatically use whatever Markdown lib and options I've configured: https://www.aleksandrhovhannisyan.com/blog/custom-markdown-components-in-11ty/

However, I inspected `this` inside the shortcode (currently getting invoked from within a blog post), and it only contains this:

```json
{
  "page": {
    "date": "2019-08-12T00:00:00.000Z",
    "inputPath": "./src/_posts/asd",
    "fileSlug": "asd",
    "filePathStem": "/_posts/asd",
    "outputFileExtension": "html",
    "url": "/blog/asd/",
    "outputPath": "_site/blog/asd/index.html"
  }
}
```

**Comment by rileymacisaac** [comment, marked as answer]
Hey @AleksandrHovhannisyan ,

Not sure if this will totally help you, but I was able to use `renderFIle` in a nunjucks shortcode that was defined in `eleventy.js`.

Example:
```
...
let renderFile = eleventyConfig.javascriptFunctions.

return renderFile('/path/to/file/file.njk', data); // data is whatever data you are passing to the file
...
```

**Reply by AleksandrHovhannisyan** [reply]
I'm trying to do something different: I want my custom JavaScript shortcode to be able to access other 11ty shortcodes (like `renderFile`).

**Reply by pdehaan** [reply]
@AleksandrHovhannisyan You should be able to have shortcodes call other shortcodes and use the EleventyRenderPlugin. I do the same thing as @rileymacisaac above where you have to grab the shortcode functions from the `eleventyConfig.javascriptFunctions` object:

```js
// .eleventy.js
const { EleventyRenderPlugin } = require("@11ty/eleventy");

module.exports = function(eleventyConfig) {
  eleventyConfig.addPlugin(EleventyRenderPlugin);

  eleventyConfig.addShortcode("footer", function (year=new Date().getFullYear()) {
    return `<footer><p>&copy; ${year}</p></footer>`;
  });

  eleventyConfig.addShortcode("bleepbloop", async function (name="") {
    const footerShortcode = eleventyConfig.javascriptFunctions.footer;
    const { renderFile, renderTemplate } = eleventyConfig.javascriptFunctions;
    return `
      ${ await renderTemplate(`## ${name}`, "md") }
      ${ await renderFile("./src/_includes/about.md") }
      ${ footerShortcode() }
    `.trim();
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

```liquid
---
# ./src/index.liquid
title: 11ty-2252
---

<h1>{{ title }}</h1>

{%- bleepbloop pkg.author -%}
```

```liquid
<!-- ./src/_includes/about.md -->
## ABOUT US

We're neat because
- fast
- cheap
- good
```

### OUTPUT
```html
<h1>11ty-2252</h1>

<h2>Peter deHaan</h2>

<h2>ABOUT US</h2>
<p>We're neat because</p>
<ul>
  <li>fast</li>
  <li>cheap</li>
  <li>good</li>
</ul>

<footer><p>&copy; 2022</p></footer>
```

So I definitely think it's possible to reuse shortcodes, although they aren't exposed in the shortcode's `this` property and require a _tiny_ bit more poking into the `eleventyConfig` object.

**Reply by pdehaan** [reply]
And because I'm super awesome, I'll add the following caveats:
- that worked beautifully with a .liquid template (because LiquidJS is rad; and promise based)
- that worked less ideally when I renamed the file to "index.njk" since Nunjucks doesn't have as rad Promise support, in fact, it rendered the following:

```html
<h1>11ty-2252</h1>

[object Promise]
```

Workaround was to convert from `addShortcode()` to `addNunjucksAsyncShortcode()`, like so:

```diff
-eleventyConfig.addShortcode("bleepbloop", async function (name="") {
+eleventyConfig.addNunjucksAsyncShortcode("bleepbloop", async function (name="") {
  const footerShortcode = eleventyConfig.javascriptFunctions.footer;
  const { renderFile, renderTemplate } = eleventyConfig.javascriptFunctions;
  return `
    ${ await renderTemplate(`## ${name}`, "md") }
    ${ await renderFile("./src/_includes/about.md") }
    ${ footerShortcode() }
  `.trim();
});

**Reply by AleksandrHovhannisyan** [reply]
Oh awesome, thank you both! This sounds like exactly what I want.

I bet I could even use a higher-order function to forward the 11ty config automatically (in my use case, my shortcodes are exported from a separate module and imported into `.eleventy.js`). So maybe something like this:

```js
// some other file
const shortcode = (eleventyConfig) => (arg) => {...}

// .eleventy.js
eleventyConfig.addShortcode('foo', shortcode(eleventyConfig));
```
